src/Host.cpp

```diff
@@ -683,49 +819,345 @@ void Host::send(QString cmd, bool wantPrint, bool dontExpandAliases)
+        // As we need to check the names we will need a different algorithm
+        // than just looking for the first gap in the integer sequence
+        // 1,2,3,...n if we want to avoid going through the list twice,
+        // the first time to check the name is not in use and a second time
+        // to find the lower positive unused integer id:
+        QMapIterator<int, stopWatch*> itStopWatch(mStopWatchMap);
+        while (itStopWatch.hasNext()) {
+            itStopWatch.next();
+            if (itStopWatch.value()->name() == name) {
+                return qMakePair(0, QStringLiteral("a stopwatch with id %1 called \"%2\" already exists").arg(QString::number(itStopWatch.key()), name));
+            }
+        }
+    }
+    int newWatchId = 1;
+    while (mStopWatchMap.contains(newWatchId)) {
+        ++newWatchId;
+    };
+
+    // It is hard to imagine a situation in which this will fail - so we won't
+    // bother coding for it:
+    auto pStopWatch = new stopWatch();
+    Q_ASSERT_X(pStopWatch, "Host::createStopWatch", "out of memory, unable to create new stopwatch");
+    if (!name.isEmpty()) {
+        pStopWatch->setName(name);
+    }
+
+    mStopWatchMap.insert(newWatchId, pStopWatch);
+    return qMakePair(newWatchId, QString());
 }
 
-double Host::getStopWatchTime(int watchID)
+bool Host::destroyStopWatch(const int id)
 {
-    if (mStopWatchMap.contains(watchID)) {
-        return static_cast<double>(mStopWatchMap[watchID].elapsed()) / 1000;
-    } else {
-        return -1.0;
+    auto pStopWatch = mStopWatchMap.take(id);
+    if (!pStopWatch) {
+        return false;
     }
+
+    delete pStopWatch;
+    return true;
 }
 
-bool Host::startStopWatch(int watchID)
+bool Host::adjustStopWatch(const int id, const qint64 milliSeconds)
 {
-    if (mStopWatchMap.contains(watchID)) {
-        mStopWatchMap[watchID].start();
+    auto pStopWatch = mStopWatchMap.value(id);
+    if (pStopWatch) {
+        pStopWatch->adjustMilliSeconds(milliSeconds);
         return true;
+    }
+
+    return false;
+}
+
+// Find the first (lowest ID) stopwatch with the given name and return its id,
+// returns 0 (not a valid id) if the name is not found. It WILL return the first
+// unnamed one if a blank string is given:
+int Host::findStopWatchId(const QString& name) const
+{
+    // Scan through existing names, in ascending id order
+    QList<int> stopWatchIdList = mStopWatchMap.keys();
+    int total = stopWatchIdList.size();
+    if (total > 1) {
+        std::sort(stopWatchIdList.begin(), stopWatchIdList.end());
+    }
+    for (int index = 0, total = stopWatchIdList.size(); index < total; ++index) {
+        auto currentId = stopWatchIdList.at(index);
+        auto pCurrentStopWatch = mStopWatchMap.value(currentId);
+        if (pCurrentStopWatch->name() == name) {
+            return currentId;
+        }
+    }
+    return 0;
+}
+
+QPair<bool, double> Host::getStopWatchTime(const int id) const
```

Comment:
Why is there no name overload of this function?

Reply:
As above.